### ISSUE TITLE
Generalize ReductionOp to support argmax and online softmax

### ISSUE THREAD

**Issue by jacobhinkle** [NVIDIA/Fuser#622]
This is a proposal to refactor reductions in nvfuser in order to support `argmin` and `argmax`, unifying them with Welford. I believe we can do this in a way that will let us implement online softmax and other efficient methods conveniently.

# Current status

We currently implement `Min` and `Max` binary operations which can appear as the reduction op in a `ReductionOp`, over any number of axes. This is what our `min` and `max` ops do, using an initial value of plus or minus infinity, respectively.

In PyTorch, `torch.amax` works like our `max`, and `torch.max` returns both the max and the argmax. Note that `torch.amax` can reduce along multiple dimensions, but `torch.max` works on either a single dimension, or on the flattened input. This is needed since it is ambiguous how one would compute a linear index when reducing multiple dimensions; is it the linear index in the entire tensor or only in the reduced dims?

# Proposed approach

I propose that we generalize `GroupedReductionOp` to `MultiReductionOp`. This would represent an arbitrary reduction over the concatenation of `M` tensors, producing `N` reduced values each of which might depend on all of the `M` tensors. All `M` tensors and `N` aggregated values can have different dtypes. Here is a sketch of the interface:
```c++
//! Describes the types of multi-reductions we could do
enum class MultiReductionBinaryOpType {
  ArgMin,
  ArgMax,
  Welford,
  Independent // simple ReductionOp or GroupedReductionOp
};

//! Describes a MultiReductionOp without the axis or input info
class MultiReductionOpType {
  MultiReductionBinaryOpType getOpType() const;

  //! How many aggregated values (output tensors) will we create?
  int getNumAggregateValues() const;
  //! Initialize the aggregate values according to the op type
  std::vector<Val*> getDefaultInitValues() const;
  std::vector<DataType>& getAggregateDataTypes() const;

  //! How many TensorViews are expected and what are their dtypes?
  int getNumTensorArgs() const;
  std::vector<DataType>& getTensorDataTypes() const;

  //! If this is an Independent, get the binary ops
  std::vector<BinaryOp> getGroupedBinaryOps() const;
  // ...
};

class MultiReductionOp : public Expr {
  MultiReductionOp(
      MultiReductionOpType op_type,
      std::vector<TensorView*> tvs,
      std::vector<int64_t> axes);
  // ...
};
```

A simple `ReductionOp` is a case where `M=N=1` and the dtypes are equal; `getGroupedBinaryOps()` would return the usual `BinaryOp`.

## Special case: argmax

We can pass two tensors to `argmax`: the input `A` and an index tensor `idx`, which we could create using `iota`. 
```c++
if (A[i] > m[i-1]) {
  m[i] = A[i];
  x[i] = idx[i];
} else {
  m[i] = m[i-1];
  x[i] = x[i-1];
}
```
This results in `m[N]` the max and `x[N]` the first corresponding entry in `idx`. We would initialize `m` with `-inf` and `x` with 0 (although this is arbitrary).

## Special case: WelfordOp

Welford can be easily expressed this way:
```
n[i] = n[i-1] + 1
avg[i] = avg[i-1] + (A[i] - avg[i-1]) / n[i]
var_sum[i] = var_sum[i-1] + (A[i] - avg[i-1]) * (A[i] - avg[i])
```
Notice that unlike `argmax` this only needs a single tensor input, so `M=1, N=3` in the notation of `MultiReductionOp`. The aggregate values have initial values (0, 0,0) and dtypes `Int, Float, Float`. Note that in order to use with rfactor, this is generalized to use a count for each A[i] instead of 1, in which case we would likely rewrite this to take `M=3` tensors, two of which are initialized to 1 and 0.

## Special case: Online softmax

In softmax, we need to convert each tensor element `A[i]` to `exp(A[i])/sum(exp(A[i]))`. A naive implementation like this quickly results in floating point overflow for large values of `A[i]`. Instead, the more stable implementation of softmax is
```
m = max(A[i])
softmax[i] = exp(A[i] - m) / sum(exp(A[i] - m))
```
This requires two successive reductions: `max` followed by `sum`.

The online softmax algorithm fuses the two reductions into a single two-way reduction:
```
m[i] = max(m[i-1], A[i])
d[i] = d[i-1] * exp(m[i-1] - m[i]) + exp(A[i] - m[i])
```
At the end of the reduction, `m[N]` is the overall max of `A` and `d[i]` is the `sum(exp(A[i] - m))` normalization term. The softmax is then formed by computing
```
softmax[i] = exp(A[i] - m[N]) / d[N]
```
The normalization term is simple to implement as an 2-way reduction op, very similar to how we would implement argmin and argmax.

If we implement this op, we could insert it in an optimization pass by pattern matching for the sumexp pattern.

For an implementation in CUDA, see https://github.com/NVIDIA/online-softmax and the manuscript [Mikalov & Gimelshein 2018](https://arxiv.org/abs/1805.02867).

## Special case: Flash attention

Flash attention takes the fusion from online softmax and goes further. In dot-product attention, softmax surrounded by matmul operations, the second of which contracts along the softmax dimension. This contraction is another reduction which flash attention fuses to the online softmax. Flash attention also does tiling, so that the first matmul and fused second matmul can operate on individual tiles; that is outside the scope of this PR. However, being able to express N-Way reductions opens the door to some day fusing them with the sum in a matmul.

See [Dao et al. 2022](https://arxiv.org/abs/2205.14135) for more info on flash attention.

# Challenges/Questions

- Like `GroupedReductionOp`, `MultiReductionOp` should take variable numbers of aggregate values and tensors. The number of needed aggregate values and tensors is determined by the op type. It may be difficult to write the CUDA helper templates without specializing for each op type.
- There is a lot of code in scheduling and lowering specializing on reductions that will require careful adaptation. We should keep that all in place while initially developing `MultiReductionOp` separately with only `argmin` and `argmax` then replacing `WelfordOp` if this approach is successful.
- Currently I believe `GroupedReductionOp` and `GroupedWelfordOp` are meant for only grid persistent reductions, while `MultiReductionOp` should work for any size and memory location. I don't know yet if there will be a challenge in supporting both with the same class. If not, we can keep the grouped versions separate indefinitely.
- Enumerating all the patterns for reduction is a bit cumbersome considering we have a rich system in place for describing arbitrary expressions already. Instead of `ArgMin` and `ArgMax`, could we instead just hold a collection of `Val`s describing outputs whose definitions relate them all to inputs which are grouped into `old_aggregates` and `new_tensor_vals`? That would let us very easily add new reduction types. For example, Welford might be defined like this:
```c++
// create constant Vals to initialize aggregates
std::vector<Val*> init_vals = {
    container->zeroVal(DataType::Int),
    container->zeroVal(DataType::Float),
    container->zeroVal(DataType::Float),
};

// create placeholder Vals for each previous aggregate value
auto n_prev = IrBuilder::create<Scalar>(DataType::Int);
auto avg_prev = IrBuilder::create<Scalar>(DataType::Float);
auto var_sum_prev = IrBuilder::create<Scalar>(DataType::Float);
std::vector<Val*> old_aggregates = {n_prev, avg_prev, var_sum_prev};

// create placeholder Vals for each new tensor input value
auto a_i = IrBuilder::create<Scalar>(DataType::Float);
std::vector<Val*> new_tensor_vals = {a_i};

// Now define the actual op using Exprs
auto n = add(n_prev, container->oneVal()); // can easily generalize to the count of a_i instead of 1
auto diff_prev = sub(a_i, avg_prev); // Note we can explicitly CSE here
auto avg = add(avg_prev, div(diff_prev, n));
auto diff_next = sub(a_i, avg);
auto var_sum = add(var_sum_prev, mul(diff_prev, diff_next));

// Output is the collection of new aggregate values
std::vector<Val*> new_aggregates = {n, avg, var_sum};

// Create the op with these flexible val lists instead of (or in addition to) an enumerated op type
IrBuilder::create<MultiReductionOp>(
    container,
    out_tensors, // one newForReduction for each of new_aggregates, with corresponding dtype
    {input_tensor},
    init_vals,
    old_aggregates,
    new_tensor_vals,
    new_aggregates);
```

```[tasklist]
### Tasks
- [ ] Implement `MultiReductionOp` with `MinWithArg` and `MaxWithArg` op types
- [ ] Update python bindings for `min` and `max` to enable access to arg versions
- [ ] Make `WelfordOp` a special case of `MultiReductionOp`
- [ ] Implement online softmax
```

**Comment by naoyam** [comment]
The idea of generalizing these non-trivial reductions is great. Thanks for sketching out the design.

I particularly like the idea of allowing reductions ops to be `Expr*`. It basically means we are expressing a lambda-like object, so this may also be a good design target for nested IR containers.

Restructuring reductions would be non-trivial. I'd recommend starting small and incrementally.

**Comment by jacobhinkle** [comment]
Note that if we also allow broadcasting in the inputs, I believe this is enough for us to enable flashattention. ![Screenshot_20230920-074614.png](https://github.com/NVIDIA/Fuser/assets/1454944/3112e320-bba3-46be-a25c-fca029d14d6a). The steps writing m_i, l_i, and O_i are a generic reduction given m_ij, l_ij, and P_ij. P is not broadcasted, but m and l are the row maxes and row sums, so they need to be broadcasted in the outer generic reduction.
